engineapi/proxy.go

```diff
@@ -172,62 +431,246 @@ type Proxy struct {
 
 	grpcClient *imclient.ProxyClient
 
-	lock *sync.RWMutex
+	lock sync.RWMutex
+
+	inUse     int
+	inUseLock sync.RWMutex
+
+	nextResync time.Time
+	resyncLock sync.RWMutex
+
+	nextGC  *time.Time
+	gcRetry int
+	gcLock  sync.RWMutex
+}
+
+type EngineClientProxy interface {
+	EngineClient
+	BackupTargetBinaryClient
+
+	getEngineClientProxy() engineClientProxy
+
+	use()
+	Done()
+}
+
+func (p *Proxy) getEngineClientProxy() engineClientProxy {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	return p
+}
+
+// use increments the number of an engine proxy client concurrent use count.
+func (p *Proxy) use() {
+	p.incrementUse(true)
+}
+
+// Done decrements the number of an engine proxy client concurrent use count.
+func (p *Proxy) Done() {
+	p.incrementUse(false)
+}
+
+type engineClientProxy interface {
+	getLogger() logrus.FieldLogger
+
+	isGRPC() bool
+	getConnectionState() connectivity.State
+
+	start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore, chan struct{}) error
+	stop() error
+
+	setNextResync(time.Duration)
+	shouldResync() bool
+
+	shouldGarbageCollect() bool
+	setNextGarbageCollect()
+	getNextGarbageCollect() *time.Time
+	getGarbageCollectRetry() int
+
+	incrementUse(bool)
+	getInUse() int
+}
+
+func (p *Proxy) getLogger() logrus.FieldLogger {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
```

Comment:
Why we need lock here? I feel this method not required, internal usage can just use logger attribute/property directly.